core/src/test/java/org/apache/iceberg/actions/TestCommitService.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.actions;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
+
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableTestBase;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Tasks;
+import org.assertj.core.api.Assertions;
+import org.awaitility.Awaitility;
+import org.junit.Test;
+
+public class TestCommitService extends TableTestBase {
+
+  public TestCommitService() {
+    super(1);
+  }
+
+  @Test
+  public void testCommittedResultsCorrectly() {
+    CustomCommitService commitService = new CustomCommitService(table, 5, 10000);
+    commitService.start();
+
+    ExecutorService executorService = Executors.newFixedThreadPool(10);
+    Tasks.range(100).executeWith(executorService).run(commitService::offer);
+    commitService.close();
+
+    Set<Integer> expected = Sets.newHashSet(IntStream.range(0, 100).iterator());
+    Set<Integer> actual = Sets.newHashSet(commitService.results());
+    Assertions.assertThat(actual).isEqualTo(expected);
+  }
+
+  @Test
+  public void testAbortFileGroupsAfterTimeout() {
+    CustomCommitService commitService = new CustomCommitService(table, 5, 200);
+    commitService.start();
+
+    // Add the number of less than rewritesPerCommit
+    for (int i = 0; i < 4; i++) {
+      commitService.offer(i);
+    }
+
+    // Simulate the latest group of rewrite
+    CustomCommitService spyCommitService = spy(commitService);
+    doReturn(false).when(spyCommitService).canCreateCommitGroup();
+    for (int i = 4; i < 8; i++) {
```

Comment:
So it took me awhile to get this.  This is to make it go above  the rewritesPerCommit limit , allowing the first batch to finally try commit right ?

Reply:
Yes, the problem could only be possible to trigger when there are more than `rewritesPerCommit ` in the queue that needs to commit by the `committerService`.  For example, we set the `rewritesPerCommit` to 5. And there are 10 file groups that need to commit by `committerService`.  The timeout occurs in the first commit (range from 0 to 4). Then main thread deletes the remaining 5 file groups (5 to 9). However, the `committerService` is not stopped and it still could commit the file groups (5 to 9).

Reply:
> However, the committerService is not stopped and it still could commit the file groups (5 to 9).

Is it true?  I dont see the group 5-9 will ever proceed because we always return false for canCreateCommitGroup().  (also ran the test to check).  Am I missing something?

I thought race condition is between commit or aborting 0-4 group.

If that is true, how about adding these comments?


```
// Add file groups 0-4 for commit.
// There are less than the rewritesPerCommit, and thus commitService will not commit until next batch added.
// During commit, these wiil sleep a fixed duration, to test race condition with close abort (below)
for (int i = 0; i < 4; i++) {
    commitService.offer(i);
 }

// Add file groups 5-9 for commit
// These are gated to not be able to commit, 
// and only serve to allow file 0-4 to proceed
CustomCommitService spyCommitService = spy(commitService);
doReturn(false).when(spyCommitService).canCreateCommitGroup();
for (int i = 4; i < 8; i++) {
   spyCommitService.offer(i);
}

// close commitService.  
// This will wait a fixed duration to abort file group 0-4
// testing race condition as they are attempt to finish commit (above)
Assertions.assertThatThrownBy(commitService::close)
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessageContaining("Timeout occurred when waiting for commits");

// Wait for the commitService to finish trying to commit file groups
Awaitility.await()
    .atMost(5, TimeUnit.SECONDS)
        .pollInSameThread()
        .untilAsserted(() -> assertThat(commitService.completedRewritesAllCommitted()).isTrue());


Assertions.assertThat(commitService.results())
      .doesNotContainAnyElementsOf(commitService.aborted);

Assertions.assertThat(commitService.results()).isEqualTo(ImmutableList.of(0, 1, 2, 3, 4));
```

Can you check if comments and my understanding is right?  

I also feel the commitService results will not always have 0-4 then?  But rather it should either be aborted or committed.

Reply:
nit: I think the original logic only has 8 groups in total.

Reply:
@szehon-ho I updated the comments, please take a look.

```
// Add file groups [0, 4) for commit.
// There are less than the rewritesPerCommit, and thus will not trigger a commit action. Those file groups
// will be added to the completedRewrites queue.
// Now the queue has 4 file groups that need to commit.
for (int i = 0; i < 4; i++) {
    commitService.offer(i);
 }

// Add file groups [4, 8) for commit
// These are gated to not be able to commit, so all those 4 file groups will be added to the queue as well.
// Now the queue has 8 file groups that need to commit.
CustomCommitService spyCommitService = spy(commitService);
doReturn(false).when(spyCommitService).canCreateCommitGroup();
for (int i = 4; i < 8; i++) {
   spyCommitService.offer(i);
}

// close commitService. 
// The committerService thread now starts to commit those remaining file groups in the completedRewrites queue.
// And also the main thread waits for the committerService thread to finish within a timeout. There are 8 file groups
// that need to commit. The first committing will commit file groups [0, 5]. This will wait a fixed duration to simulate
// the time cost exceeding the timeout. 
// Then testing race conditions between the main thread (trying to abort remaining file groups[6, 8)), and the 
// committerService thread (trying to commit file groups[6, 8))
Assertions.assertThatThrownBy(commitService::close)
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessageContaining("Timeout occurred when waiting for commits");

// Wait for the commitService to finish. Committed all file groups or abort remaining file groups. 
Awaitility.await()
    .atMost(5, TimeUnit.SECONDS)
        .pollInSameThread()
        .untilAsserted(() -> assertThat(commitService.completedRewritesAllCommitted()).isTrue());


Assertions.assertThat(commitService.results())
      .doesNotContainAnyElementsOf(commitService.aborted);

Assertions.assertThat(commitService.results()).isEqualTo(ImmutableList.of(0, 1, 2, 3, 4));
```
> I also feel the commitService results will not always have 0-4 then? But rather we should change the check to if they are either aborted or committed.

[0, 4] should be committed success, however, the [5, 8] may be aborted or committed. I think we need to keep the following:
```
Assertions.assertThat(commitService.results())
      .doesNotContainAnyElementsOf(commitService.aborted);
```
This indicates the committed and aborted should not be overlapped.

Reply:
Thanks, it makes the test easier to follow now.   Now i get, that when close, there's an extra condition in the main commitService loop to commit those.

To me,  this is the most confusing part.
```
doReturn(false).when(spyCommitService).canCreateCommitGroup();
```
I thought this would prevent group 4-8 from ever proceeding, but I guess its just to prevent the main thread from commiting immediately.  It still gets on the list, and background thread can commit whenever the condition is triggered?

Reply:
Suggestion to make it a bit shorter.

```
// close commitService. 
// The committerService thread now starts to commit those remaining file groups in the completedRewrites queue.
// And also the main thread waits for the committerService thread to finish within a timeout. There are 8 file groups
// that need to commit. The first committing will commit file groups [0, 5]. This will wait a fixed duration to simulate
// the time cost exceeding the timeout. 
// Then testing race conditions between the main thread (trying to abort remaining file groups[6, 8)), and the 
// committerService thread (trying to commit file groups[6, 8))
```
=>
```
// close commitService. 
// This allows committerService thread to starts to commit the remaining file groups [0-8] in the completedRewrites queue.
// And also the main thread waits for the committerService thread to finish within a timeout. 

// The committerService thread commits file groups [0, 5].  These will wait a fixed duration to simulate timeout on the main thread, which then tries to abort file groups [6-8].
// This tests the race conditions, as the commiterService is also trying to commit groups [6-8].
```

Reply:
> but I guess its just to prevent the main thread from commiting immediately. 

Actually this is prevent the committing from the rewrite thread pool.

> and committerService thread can commit whenever the condition is triggered as its not using the spy?

That's right.